src/libraries/System.Private.CoreLib/src/System/Math.DivModInt.cs

```diff
@@ -0,0 +1,234 @@
+
+        [StackTraceHidden]
+        internal static ulong DivUInt64(ulong dividend, ulong divisor)
+        {
+            if (Hi32Bits(divisor) == 0)
+            {
+                if ((uint)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if (Hi32Bits(dividend) == 0)
+                    return (uint)dividend / (uint)divisor;
+            }
+
+            return DivUInt64Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static int ModInt32(int dividend, int divisor)
+        {
+            if ((uint)(divisor + 1) <= 1) // Unsigned test for divisor in [-1 .. 0]
+            {
+                if (divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+                else if (divisor == -1)
+                {
+                    if (dividend == int.MinValue)
+                    {
+                        ThrowHelper.ThrowOverflowException();
+                    }
+                    return 0;
+                }
+            }
+
+            return ModInt32Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static uint ModUInt32(uint dividend, uint divisor)
+        {
+            if (divisor == 0)
+            {
+                ThrowHelper.ThrowDivideByZeroException();
+            }
+
+            return ModUInt32Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static long ModInt64(long dividend, long divisor)
+        {
+            if (Is32BitSigned(divisor))
+            {
+                if ((int)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if ((int)divisor == -1)
+                {
+                    if ((ulong)dividend == 0x8000000000000000ul)
+                    {
+                        ThrowHelper.ThrowOverflowException();
+                    }
+                    return 0;
+                }
+
+                if (Is32BitSigned(dividend))
+                {
+                    return (int)dividend % (int)divisor;
+                }
+            }
+
+            return ModInt64Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static ulong ModUInt64(ulong dividend, ulong divisor)
+        {
+            if (Hi32Bits(divisor) == 0)
+            {
+                if ((uint)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if (Hi32Bits(dividend) == 0)
+                    return (uint)dividend % (uint)divisor;
+            }
+
+            return ModUInt64Internal(dividend, divisor);
+        }
+
+        // helper method to get high 32-bit of 64-bit int
+        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
```

Comment:
Why do we need AggressiveOptimization on this helper when it is marked with AggressiveInlining?

Reply:
The tier 0 code (-c Release) looked like:

```asm
G_M000_IG01:                ;; offset=0x0000
       push     ebp
       mov      ebp, esp
       sub      esp, 12
 
G_M000_IG02:                ;; offset=0x0006
       push     dword ptr [ebp+0x0C]
       push     dword ptr [ebp+0x08]
       call     [Sgstem.Math:Hi32Bits(long):uint]
       mov      dword ptr [ebp-0x04], eax
       push     dword ptr [ebp+0x0C]
       push     dword ptr [ebp+0x08]
       call     [System.Int32:CreateTruncating[long](long):int]
       mov      ecx, eax
       call     [System.Int64:CreateTruncating[int](int):long]
       mov      dword ptr [ebp-0x0C], eax
       mov      dword ptr [ebp-0x08], edx
       push     dword ptr [ebp-0x08]
       push     dword ptr [ebp-0x0C]
       call     [Sgstem.Math:Hi32Bits(long):uint]
       cmp      eax, dword ptr [ebp-0x04]
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x0044
       mov      esp, ebp
       pop      ebp
       ret      8
 
; Total bytes of code 74
```

now it's FullOpts:

```asm
G_M000_IG01:                ;; offset=0x0000
 
G_M000_IG02:                ;; offset=0x0000
       mov      eax, dword ptr [esp+0x08]
       mov      edx, dword ptr [esp+0x04]
       sar      edx, 31
       cmp      eax, edx
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x0013
       ret      8
 
; Total bytes of code 22
```

I think #110267 series of PRs will fix this issue when helpers are inlined, then we won't need these flags.

Reply:
The tier 0 code (-c Release) looked like:

```asm
G_M000_IG01:                ;; offset=0x0000
       push     ebp
       mov      ebp, esp
       sub      esp, 12
 
G_M000_IG02:                ;; offset=0x0006
       push     dword ptr [ebp+0x0C]
       push     dword ptr [ebp+0x08]
       call     [System.Math:Hi32Bits(long):uint]
       mov      dword ptr [ebp-0x04], eax
       push     dword ptr [ebp+0x0C]
       push     dword ptr [ebp+0x08]
       call     [System.Int32:CreateTruncating[long](long):int]
       mov      ecx, eax
       call     [System.Int64:CreateTruncating[int](int):long]
       mov      dword ptr [ebp-0x0C], eax
       mov      dword ptr [ebp-0x08], edx
       push     dword ptr [ebp-0x08]
       push     dword ptr [ebp-0x0C]
       call     [System.Math:Hi32Bits(long):uint]
       cmp      eax, dword ptr [ebp-0x04]
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x0044
       mov      esp, ebp
       pop      ebp
       ret      8
 
; Total bytes of code 74
```

now it's FullOpts:

```asm
G_M000_IG01:                ;; offset=0x0000
 
G_M000_IG02:                ;; offset=0x0000
       mov      eax, dword ptr [esp+0x08]
       mov      edx, dword ptr [esp+0x04]
       sar      edx, 31
       cmp      eax, edx
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x0013
       ret      8
 
; Total bytes of code 22
```

I think #110267 series of PRs will fix this issue when helpers are inlined, then we won't need these flags.

Reply:
It is normal that Tier-0 code is very inefficient.

In default config, we should get R2R code for the helper as Tier-0. The R2R should be reasonably optimized code.

Reply:
Also, I think it would be better for #110267 series of change to avoid calling more helpers from the aggresively inlined helper. Things like that tend to eat inlining budget pretty fast.

Ie change `if (Hi32Bits(divisor) == 0)` to `if ((uint)((ulong)divisor >> 32) == 0)` and so on like you have done in some places.

Reply:
I will inline the non-CreateTruncating version, basically this code without separate methods: https://github.com/dotnet/runtime/commit/40dba4f7613d2bcbfbebd65455dabca9ba74df74. Its summary is:

With AggressiveInlining|AggressiveOptimization
```asm
G_M000_IG01:                ;; offset=0x0000
 
G_M000_IG02:                ;; offset=0x0000
       mov      eax, dword ptr [esp+0x04]
       sar      eax, 31
       cmp      dword ptr [esp+0x08], eax
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x0011
       ret      8
 
; Total bytes of code 20
```

With AggressiveInlining
```asm
G_M000_IG01:                ;; offset=0x0000
       push     ebp
       mov      ebp, esp
       push     eax
 
G_M000_IG02:                ;; offset=0x0004
       mov      eax, dword ptr [ebp+0x0C]
       xor      edx, edx
       mov      edx, dword ptr [ebp+0x08]
       mov      dword ptr [ebp-0x04], edx
       mov      edx, dword ptr [ebp-0x04]
       sar      edx, 31
       xor      ecx, ecx
       cmp      eax, edx
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x001F
       mov      esp, ebp
       pop      ebp
       ret      8
 
; Total bytes of code 37
```
Without attrubte
```asm
G_M000_IG01:                ;; offset=0x0000
       push     ebp
       mov      ebp, esp
       push     eax
 
G_M000_IG02:                ;; offset=0x0004
       mov      eax, dword ptr [ebp+0x0C]
       xor      edx, edx
       mov      edx, dword ptr [ebp+0x08]
       mov      dword ptr [ebp-0x04], edx
       mov      edx, dword ptr [ebp-0x04]
       sar      edx, 31
       xor      ecx, ecx
       cmp      eax, edx
       sete     al
       movzx    eax, al
 
G_M000_IG03:                ;; offset=0x001F
       mov      esp, ebp
       pop      ebp
       ret      8
 
; Total bytes of code 37
```

Reply:
Is this comparing Tier0 and Tier1 code?

AggressiveOptimization disables R2R and tiered compilation for the method and makes the JIT generate fully optimized method right away. AggressiveOptimization is bad for startup time.